### PR TITLE
4bit quantization with bitsandbytes

### PR DESCRIPTION
**Pull request by michaelfeil**
Restored from 27.07.2023, dev-branch.

### CODE REVIEW

**Comment by Atry** [comment]
I tried to pull this PR to our internal repository and run CI, the tests fail on CUDA but succeeded on CPU.

Could you run unit tests and paste the test log here?

File: server/text_generation_server/utils/layers.py [hunk 1]
```diff
@@ -152,9 +185,21 @@ def get_linear(weight, bias, quantize):
...
+    elif quantize == "bitsandbytes-nf4":
+        linear = Linear4bit(
+            weight,
+            bias,
+            quant_type="nf4",
+        )
     elif quantize == "gptq":
         try:
-            qweight, qzeros, scales, g_idx, bits, groupsize, use_exllama = weight
+            qweight, qzeros, scales, g_idx, bits, groupsize = weight
```
[the remark this answers is not in the document]
Comment (michaelfeil) [reply server/text_generation_server/utils/layers.py]: You are correct. Should be fixed now.

**Comment by michaelfeil** [comment]
@atry Can you give more details. I am pretty confident, the merge request can be merged like this.

**Comment by Atry** [comment]
Thank you for the quick response. Let me check if it's due to our internal code base incompatibility.

File: server/text_generation_server/utils/layers.py [hunk 2]
```diff
@@ -219,31 +264,36 @@ def load(config, prefix: str, weights):
...
+                world_out, gather_input, group=self.process_group
             )
-            return world_out
 
-        output = super().forward(input)
-        if not self.should_gather:
-            return output
+            if input.shape[0] == 1:
+                return world_out
+            return world_out.T
```
Comment (Atry) [inline server/text_generation_server/utils/layers.py:291]: I don't understand this. Why do we return `world_out.T`?

File: server/text_generation_server/utils/layers.py [hunk 3]
```diff
@@ -140,6 +140,39 @@ def forward(self, x: torch.Tensor):
...
+
+    def forward(self, x: torch.Tensor):
+        # weights are cast automatically as Int8Params, but the bias has to be cast manually
+        if self.bias is not None and self.bias.dtype != x.dtype:
+            self.bias.data = self.bias.data.to(x.dtype)
+
+        if getattr(self.weight, "quant_state", None) is None:
+            print(
+                "FP4 quantization state not initialized. Please call .cuda() or .to(device) on the LinearFP4 layer first."
+            )
```
Comment (Atry) [inline server/text_generation_server/utils/layers.py:161]: Why is `print` used here? Shall we raise an error here?

File: server/text_generation_server/utils/layers.py [hunk 4]
```diff
@@ -140,6 +140,39 @@ def forward(self, x: torch.Tensor):
...
+            weight.data, requires_grad=False, compress_statistics=True, quant_type=quant_type
+        )
+        self.compute_dtype = None
+        self.weight.cuda(weight.device)
+        self.bias = bias
+
+    def forward(self, x: torch.Tensor):
+        # weights are cast automatically as Int8Params, but the bias has to be cast manually
+        if self.bias is not None and self.bias.dtype != x.dtype:
+            self.bias.data = self.bias.data.to(x.dtype)
```
Comment (Atry) [inline server/text_generation_server/utils/layers.py:156]: Why do we need to cast `bias` in `forward`? Should we do it in `__init__` instead?

**Comment by michaelfeil** [comment]
I am unsure about the implementation, this is restored from 27.07.2023, dev-branch. I have personally not implemented these changes, just merging the existing branch from the fork.

**Comment by Atry** [comment]
I got the following error when running tests. Are you familiar with it?
```
thread 'validation::tests::test_validation_input_length' panicked at 'called `Result::unwrap()` on an `Err` value: Error("EOF while parsing a value", line: 1, column: 0)', router/src/lib.rs
```

**Comment by Atry** [comment]
I fixed the compatibility issue between this PR and our internal fork. It seems work. Now merging.

**Pull request merged**